### ISSUE TITLE
ENH: Saliency Maps

### ISSUE THREAD

**Issue by eaguaida** [shap/shap#3844]
### Problem Description

I've realized that SHAP doesn't have an implementation for Saliency Maps. Even though image_plots lets us compute the Shapley values for an image, it doesn't provide an attention map.

By creating this implementation, SHAP can be compared against other methods like Grad-CAM or RISE, which SHAP outperforms anyway. At the same time, this saliency map can be benchmarked using Causal Metrics as described in https://arxiv.org/pdf/2311.17876.

### Feature request checklist

- [X] I have checked the issue tracker for duplicate issues.
- [X] I'd be interested in making a PR to implement this feature

**Comment by connortann** [comment]
Thanks for creating the issue and the PR! It's great to have new contributors on board.

Please could you provide a bit more detail on the differences between `image_plot` and the proposed new `saliency_plot`? Are they both fundamentally a grayscale image with `ax.imshow(shap_values)` superposed on top?

It would help if you were able to provide an example of both plots for the same `Explanation` object

**Comment by eaguaida** [comment]
Hey Connor! You are welcome, the pleasure is mine :)

While they are both grayscale with values superposed on top, my method extracts the values from the heatmap and uses bilinear interpolation to produce an attention map

I was doing some experimentation and comparisons between RISE[[1]](https://arxiv.org/pdf/1806.07421) and Grad-CAM[[2]](https://arxiv.org/pdf/1610.02391) as they both display a Saliency Map, and thought that this could be a nice and complementary feature for SHAP to have

For example, in Christopher Molnar book about Interpretable ML[[3]](https://christophm.github.io/interpretable-ml-book/pixel-attribution.html), SHAP is not mentioned in the Pixel Attribution (Saliency Maps) section

A few papers [[5]](https://arxiv.org/pdf/1806.07421), [[6]](https://www.researchgate.net/publication/358261598_Metrics_for_saliency_map_evaluation_of_deep_learning_explanation_methods) discuss about creating techniques to compare these Saliency Maps, and I've seen Saliency Maps mainly being used in Medical Settings [[5]](https://pubs.rsna.org/doi/10.1148/ryai.2021200267),[[6]](https://www.nature.com/articles/s42256-022-00536-x),[[7]](https://www.researchgate.net/figure/Representative-original-images-and-saliency-maps-of-WA-F-and-AM-F-The-regions-of_fig5_351914648)

Anyway, just let me know your thoughts! I've also pushed a notebook alongside some images in my branch to test the functions out. I am happy to write the doc section explaining everything in detail 

![shap_image_plot](https://github.com/user-attachments/assets/5b662129-fda4-4f28-b55d-31e176acce48)
![shap_saliency_maps](https://github.com/user-attachments/assets/23ed8ea4-4dcb-45b4-936a-3eae8b9f6b13)

**Comment by connortann** [comment]
Thanks for the explanation @eaguaida !

## Defining saliency

I'm a bit unclear on how the saliency is defined and whether that makes intuitive sense. It looks like the SHAP values are normalised to between 0-1, for each image and for each class. This seems to have a few consequences:

1. The treatment of any large negative SHAP values doesn't quite make intuitive sense. Arguably this is where the model is paying "attention", so perhaps the saliency should actually be _high_ for these regions rather than low?
2. The saliency maps for irrelevant classes look quite strange, such as for the bottom-right example of a `black-footed_ferret`. The SHAP values are close to zero, and so the normalised map has some exaggerated artifacts which are presumably quite meaningless. Would it be more useful to normalise across all classes, so that these small SHAP values are not exaggerated?
3. As a side note, we should probably have a different default color scale for saliency, as Red & Blue are used throughout the package to represent positive and negative SHAP values.


## Implementation

Thinking out loud, it strikes me that the two types of plot are similar, so I wonder if the same functionality could be added to `image_plot` by the use of some extra arguments. For example, is this new functionality equivalent to:
1. an option for smoothing (Bilinear interpolation), and
2. an option to normalise values in each image to the range 0-1?

@thatlittleboy  @CloseChoice , what do you think?

**Comment by eaguaida** [comment]
Thanks for the review! 

> ## Defining saliency
> I'm a bit unclear on how the saliency is defined and whether that makes intuitive sense. It looks like the SHAP values are normalised to between 0-1, for each image and for each class. This seems to have a few consequences:

> 
> 1. The treatment of any large negative SHAP values doesn't quite make intuitive sense. Arguably this is where the model is paying "attention", so perhaps the saliency should actually be _high_ for these regions rather than low?


Yes, I think the normalization I did there was pointless. It was meant to be used for comparison between other saliency map algorithms normalized between 0 to 1, hence why I removed the bar just to show the POC. But now I'm seeing more clearly

I don't know if it's in my head, but it seems to me that there's a terminology division in the field, maybe? Some consider the negative values as "most ignored ones," meaning that the model didn't bother to look at them since they are not relevant to the class. Therefore, a normalization between 0 to 1 would make sense if we are only interested in looking for the "relevant" values, inclining more with the concept of "Attention Map"

I totally understand that in SHAP, the negative values have "gained attention of the model" in a way, since it's saying: "yeah, this subset is definitely not part of the class" and influences the decision negatively. Are both concepts okay in different ways, or is it me who is being confused? I'm just trying to understand more about how to plot the saliency in a more descriptive way

Would the example below be more aligned?

![lime_rise_saliency](https://github.com/user-attachments/assets/145fa0a6-c5d4-440c-bca2-0717d2357c65)

> 2. The saliency maps for irrelevant classes look quite strange, such as for the bottom-right example of a `black-footed_ferret`. The SHAP values are close to zero, and so the normalised map has some exaggerated artifacts which are presumably quite meaningless. Would it be more useful to normalise across all classes, so that these small SHAP values are not exaggerated?

I see the error now, you are right! Normalising in one class doesn't really highlight well for irrelevant classes

> 3. As a side note, we should probably have a different default color scale for saliency, as Red & Blue are used throughout the package to represent positive and negative SHAP values.

Sure! I tried the 'red_transparent_blue' one but it didn't look good to me. I think a custom one could create a better distinction

**Comment by connortann** [comment]
> I totally understand that in SHAP, the negative values have "gained attention of the model" in a way, since it's saying: "yeah, this subset is definitely not part of the class" and influences the decision negatively. Are both concepts okay in different ways?

Yes, I think both ways of displaying the values are okay and could be used depending on the context. I agree it would be helpful to have the capability to create plots that look more similar to "attention maps" from other methods.

Rather than adding this as a separate function, how do you feel about extending the existing `image_plot` function with some extra options to allow this kind of customisation? 

You can already customise the colormap, and for this case you'd presumably want to use a [sequential](https://matplotlib.org/stable/users/explain/colors/colormaps.html#sequential) colormap like `cmap="Blues"` that only emphasise high SHAP values.

So I think we just need two things:
1. An option for bilinear interpolation
2. An option to normalise (nb. this needs a bit more discussion)

If we added those options, would that satisfy your use case?

**Comment by CloseChoice** [comment]
@eaguaida thanks for having a look at this. I think it would be great to have this in shap but I am a bit skeptical of the name. Aren't saliency maps something where we calculate the gradient of the probability of the class regarding the inputs? If so then I think we could tweak the name somehow to make sure that we are doing something similar but tweaked to shap values.

Regarding the negative values: I interpret those in this scenario that the model strongly believes that the corresponding pixels do not belong to the class.
I would also support the possibility to have colormaps but wouldn't be too afraid to set blue-red as a default since I believe most users will prefer

**Comment by eaguaida** [comment]
Sorry for the delay, it's been a hectic week! 


> 
> Regarding the negative values: I interpret those in this scenario that the model strongly believes that the corresponding pixels do not belong to the class. I would also support the possibility to have colormaps but wouldn't be too afraid to set blue-red as a default since I believe most users will prefer

@CloseChoice 

From my understanding, terms like saliency maps/attribution map are interchangeable, when we calculate the gradient of the probability of the class regarding the inputs that's what Grad-CAM is doing, I see it more like a virus going inside of the model perturbating it, which then would be a white-box method. But there are few others algorithms that do the same and create a saliency map by masking pixels and measuring the inference prediction, by using an analogy we can say that this is the model "psychologist", so the saliency map will barely be that heatmap visual representation of where a model is "paying attention" 

Maybe we can rephrase it to say that saliency map/attention map is that where the negative values are set to 0, and we only count for positive values? - Let me know what you think
 
Some examples, https://arxiv.org/abs/2311.14471 - Blake states that a disadvantage is that "any white-box method is the necessity of having access to the internals of the model. This may not be available depending upon the nature of various healthcare and commercial collaborations, and even between purely academic teams there may be a reluctance to share"

RISE also uses the word "Saliency Map" and is only measuring predictions/pixels covered so doesn't go into the model internal gradients https://arxiv.org/pdf/1806.07421

https://christophm.github.io/interpretable-ml-book/pixel-attribution.html

Taking it from here, we can say that SHAP would be a Perturbation-based method (like RISE) and Grad-CAM a Gradient-based method, the fact that gradient-based methods are commonly associated with saliency maps is just they were the first ones to output a heatmap like that. RISE creates a Saliency Map by using bilinear interpolation in the masks used to perturbate the input, CAM generates a heatmap full of squares (like image_plot in shap) and then somehow convert it into a saliency map (I haven't dig much into it, just experiment inside of it and saw that, and thought that SHAP could do it as well). That's my thought process of why with upsampling methods we can generate a saliency map easier. We can just test it and classify it as a visualization trick to generate saliency maps

However, I also have a few notebooks using techniques for comparing Saliency Map with SHAP, RISE and Grad-CAM, and SHAP achieve higher results than all of them. Based on that I don't see why this implementation cannot be seen as a Saliency Map if it's even performing better

Please feel free to let me know if I am confusing you more or if what I am saying it's actually making sense haha

Don't get me wrong, I'm not pushy about the name but just thought that "saliency maps" would be beneficial in terms of SEO or keywords, if you search "Shap Saliency Maps" there's no much information, so it might be just a problem of rebranding, also from my work experience, in research labs SHAP is still being tested against those saliency maps methods even if the outputs are different

> > 
> 
> 1. An option for bilinear interpolation
> 

@connortann 

The bilinear option seems great to me, would you recommend me to code it from scratch? or just use Torchvision instead? It would be ideal to use a small subset of options like OpenCV with:
![image](https://github.com/user-attachments/assets/9b8516b0-5f31-47ca-939e-d380ee422a91)

We can just start with some flags like: 

1. `'interpolation' = bilinear,` -> bilinear interpolation for more localised or smaller explanations,  
2. `'interpolation' = area,` for upsampling using pixel area relation

Let me know you thoughts!

**Comment by CloseChoice** [comment]
@eaguaida Thanks for the long and thoughtful answer. That and a bit of reading changed my mind. I think I overestimated how much of a fixed term "saliency map" actually is. Together with the key-word argument `shap_values` should make it pretty clear how we generate the plot. So from my point of view everything is sorted.

**Comment by eaguaida** [comment]
Just edited my comment since there was something duplicated, thanks for reading it through!
